src/CrawlQueue/ArrayCrawlQueue.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Spatie\Crawler\CrawlQueue;
+
+use Psr\Http\Message\UriInterface;
+use Spatie\Crawler\CrawlUrl;
+use Spatie\Crawler\Exception\UrlNotFoundByIndex;
+use TypeError;
+
+/**
+ * Crawl queue implemented with arrays.
+ */
+class ArrayCrawlQueue implements CrawlQueue
+{
+    /**
+     * All known URLs, indexed by URL string.
+     *
+     * @var CrawlUrl[]
+     */
+    protected $urls = [];
+
+    /**
+     * Pending URLs, indexed by URL string.
+     *
+     * @var CrawlUrl[]
+     */
+    protected $pendingUrls = [];
+
+    public function add(CrawlUrl $url) : CrawlQueue
+    {
+        $urlString = (string) $url->url;
+
+        if (! isset($this->urls[$urlString])) {
+            $url->setId($urlString);
+
+            $this->urls[$urlString] = $url;
+            $this->pendingUrls[$urlString] = $url;
+        }
+
+        return $this;
+    }
+
+    public function hasPendingUrls() : bool
+    {
+        return (bool) $this->pendingUrls;
+    }
+
+    public function getUrlById($id) : CrawlUrl
+    {
+        if (! isset($this->urls[$id])) {
+            throw new UrlNotFoundByIndex("Crawl url $id not found in collection.");
+        }
+
+        return $this->urls[$id];
+    }
+
+    public function hasAlreadyBeenProcessed(CrawlUrl $url) : bool
+    {
+        $url = (string) $url->url;
+
+        return ! isset($this->pendingUrls[$url]) && isset($this->pendingUrls[$url]);
+    }
+
+    public function markAsProcessed(CrawlUrl $crawlUrl)
+    {
+        $url = (string) $crawlUrl->url;
+
```

Comment:
- add ``$url`` to ``$this->processedUrls``

Reply:
@AbdelkaderBah could you PR your proposals (with tests) to master?

Reply:
Thanks for adding some tests and for releasing, @freekmurze! 👍

Note that I'm not fond of adding an `InvalidUrl` exception: this is not an exception that's meant to be caught in userland, it can only happen if there is a bug in the code. For this reason, I think using built-in exceptions (`InvalidArgumentException`, or better yet in PHP 7, `TypeError` for consistency with PHP's built-in type checker) is the way to go.

If the justification is for moving the exception message logic out of the main code flow, I think a separate factory method for the exception would be appropriate.

Anyway it's released now, let's discuss this later in v5!